packages/hardhat/contracts/YourCollectible.sol

```diff
@@ -39,4 +121,35 @@ contract YourCollectible is ERC721, Groth16Verifier {
 		require(_tokenId < _nextTokenId, "tokenId does not exist");
 		return "https://austingriffith.com/images/paintings/buffalo.jpg";
 	}
+
+	// ----------------------------------------------------------
+	// Utility functions for destructuring a proof publicSignals|
+	// ----------------------------------------------------------
+
+	function getWaterMarkFromPublicSignals(
+		uint256[38] memory _pubSignals
+	) public pure returns (uint256) {
+		return _pubSignals[37];
+	}
+
+	// Numbers of events is arbitary but for this example we are using 10 (including test eventID)
+	function getValidEventIdFromPublicSignals(
+		uint256[38] memory _pubSignals
+	) public pure returns (uint256[10] memory) {
+		// Events are stored from starting index 15 to next 10 indexes
+		uint256[10] memory eventIds;
+		for (uint256 i = 0; i < 10; i++) {
+			eventIds[i] = _pubSignals[15 + i];
+		}
+		return eventIds;
+	}
```

Comment:
Rather than hardcoding in 10, we should calculate this size dynamically. The way we can do this is that the length of publicSignals depends on the number of valid event IDs. So I believe you can just subtract a constant (I believe 18?) from _pubSignals.length

Reply:
Tysm Richard 🙌 

Just a doubt :  

> The way we can do this is that the length of publicSignals depends on the number of valid event IDs 

The length of `pubSignals` is constant is right i.e `38`? 

Looking at the way we construct `pubSignals` at frontend : 

1. Till index `0-14` we push other metadata like `ticketId` , `productId` etc 
2. And then we fill the next 20 indexes with `EventIds`  i.e from `15-34`
    - If the event's length is greater than 20 we throw error [here](https://github.com/BuidlGuidl/devconnect-zupass-se2/blob/feda84193916384542200349953291bf1de93ae4/packages/nextjs/utils/scaffold-eth/pcd.ts#L35)
    - If its less than 20 then we pad it with [`BABY_JUB_NEGATIVE_ONE` variable](https://github.com/BuidlGuidl/devconnect-zupass-se2/blob/feda84193916384542200349953291bf1de93ae4/packages/nextjs/utils/scaffold-eth/pcd.ts#L45)
3. Last three indexes with `watermark` etc

>  So I believe you can just subtract a constant (I believe 18?) from _pubSignals.length

Didn't get this 😅 

But in the latest commit updated it to be more dynamic if that makes sense, but 100% agree that it should be dynamic instead of hardcode

But please let me know if I missed something will update it, Tysm 🙌

Reply:
Ah, but I think the right function here would be `publicSignalsFromClaim`, not `snarkInputForValidEventIds`. 

We end up passing in _pubSignals [here](https://github.com/BuidlGuidl/devconnect-zupass-se2/blob/feda84193916384542200349953291bf1de93ae4/packages/nextjs/utils/scaffold-eth/pcd.ts#L96)

Reply:
Ah, I think you're right here. That makes sense!

Reply:
One request - any might be wrong since it's solidity-related 😅 - could we remove this length constant and just append to the array, stopping once we either (1) hit a value that is BABY_JUB_NEGATIVE_ONE or (2) get to length 20. That way, we won't have this other variable VALID_EVENT_IDS_LENGTH that we might forget to update during the demo.

Reply:
Ah my bad - was looking at the wrong commit and saw you just used `.length`. That's actually way cleaner, thanks!